File: src/rust/geometer-client/src/indexed_mesh_packet_a0.rs
//! Canonical encoder for `geometry.indexed_triangle_mesh.packet.a0`.

const MAGIC: &[u8; 8] = b"GMIMSH01";
const HEADER_BYTES: usize = 64;
const HAS_SOURCE_FACES: u32 = 1;
const MAX_PACKET_BYTES: usize = 268_435_456;
const MAX_VERTICES: usize = 2_000_000;
const MAX_TRIANGLES: usize = 2_000_000;
const UNSPECIFIED_SOURCE_FACE: u32 = u32::MAX;

pub const INDEXED_TRIANGLE_MESH_MEDIA_TYPE: &str =
    "application/vnd.wavenumber.geometer.indexed-triangle-mesh";
pub const INDEXED_TRIANGLE_MESH_PACKET_FORMAT: &str = "geometry.indexed_triangle_mesh.packet.a0";

#[derive(Clone, Debug, PartialEq)]
pub struct IndexedTriangleMeshA0 {
    pub positions: Vec<[f64; 3]>,
    pub triangles: Vec<[u32; 3]>,
    pub source_faces: Option<Vec<u32>>,
}

#[derive(Clone, Debug, PartialEq, Eq, thiserror::Error)]
pub enum IndexedMeshPacketError {
    #[error("indexed mesh is empty or exceeds its governed limits")]
    InvalidCount,
    #[error("indexed mesh contains a non-finite position")]
    NonFinitePosition,
    #[error("indexed mesh triangle contains an invalid vertex index")]
    InvalidTriangle,
    #[error("source_faces must contain exactly one value per triangle")]
    InvalidSourceFaces,
    #[error("indexed-mesh packet exceeds 268 MiB")]
    PacketTooLarge,
}

pub fn encode_indexed_triangle_mesh_a0_packet(
    mesh: &IndexedTriangleMeshA0,
) -> Result<Vec<u8>, IndexedMeshPacketError> {
    validate_mesh(mesh)?;
    let layout = PacketLayout::for_mesh(mesh);
    if layout.packet_bytes > MAX_PACKET_BYTES {
        return Err(IndexedMeshPacketError::PacketTooLarge);
    }

    let mut output = vec![0_u8; layout.packet_bytes];
    write_header(&mut output, mesh, layout);
    write_payload(&mut output, mesh, layout);
    Ok(output)
}

fn validate_mesh(mesh: &IndexedTriangleMeshA0) -> Result<(), IndexedMeshPacketError> {
    validate_counts(mesh)?;
    validate_positions(mesh)?;
    validate_triangles(mesh)?;
    validate_source_faces(mesh)
}

fn validate_counts(mesh: &IndexedTriangleMeshA0) -> Result<(), IndexedMeshPacketError> {
    if mesh.positions.is_empty()
        || mesh.triangles.is_empty()
        || mesh.positions.len() > MAX_VERTICES
        || mesh.triangles.len() > MAX_TRIANGLES
    {
        return Err(IndexedMeshPacketError::InvalidCount);
    }
    Ok(())
}

fn validate_positions(mesh: &IndexedTriangleMeshA0) -> Result<(), IndexedMeshPacketError> {
    if mesh
        .positions
        .iter()
        .flatten()
        .any(|value| !value.is_finite())
    {
        return Err(IndexedMeshPacketError::NonFinitePosition);
    }
    Ok(())
}

fn validate_triangles(mesh: &IndexedTriangleMeshA0) -> Result<(), IndexedMeshPacketError> {
    let vertex_count = mesh.positions.len();
    if mesh.triangles.iter().any(|triangle| {
        triangle.iter().any(|index| *index as usize >= vertex_count)
            || triangle[0] == triangle[1]
            || triangle[1] == triangle[2]
            || triangle[2] == triangle[0]
    }) {
        return Err(IndexedMeshPacketError::InvalidTriangle);
    }
    Ok(())
}

fn validate_source_faces(mesh: &IndexedTriangleMeshA0) -> Result<(), IndexedMeshPacketError> {
    if mesh
        .source_faces
        .as_ref()
        .is_some_and(|values| values.len() != mesh.triangles.len())
    {
        return Err(IndexedMeshPacketError::InvalidSourceFaces);
    }
    Ok(())
}

fn has_source_faces(mesh: &IndexedTriangleMeshA0) -> bool {
    mesh.source_faces
        .as_ref()
        .is_some_and(|values| values.iter().any(|value| *value != UNSPECIFIED_SOURCE_FACE))
}

#[derive(Clone, Copy)]
struct PacketLayout {
    positions_offset: usize,
    triangles_offset: usize,
    source_faces_offset: usize,
    packet_bytes: usize,
}

impl PacketLayout {
    fn for_mesh(mesh: &IndexedTriangleMeshA0) -> Self {
        let positions_offset = HEADER_BYTES;
        let triangles_offset = positions_offset + mesh.positions.len() * 24;
        let triangles_end = triangles_offset + mesh.triangles.len() * 12;
        let source_faces_offset = if has_source_faces(mesh) {
            align_eight(triangles_end)
        } else {
            0
        };
        let payload_end = if source_faces_offset == 0 {
            triangles_end
        } else {
            source_faces_offset + mesh.triangles.len() * 4
        };
        Self {
            positions_offset,
            triangles_offset,
            source_faces_offset,
            packet_bytes: align_eight(payload_end),
        }
    }
}

fn write_header(output: &mut [u8], mesh: &IndexedTriangleMeshA0, layout: PacketLayout) {
    output[..8].copy_from_slice(MAGIC);
    put_u16(output, 8, 1);
    put_u16(output, 10, HEADER_BYTES as u16);
    let flags = if layout.source_faces_offset == 0 {
        0
    } else {
        HAS_SOURCE_FACES
    };
    put_u32(output, 12, flags);
    put_u64(output, 16, layout.packet_bytes as u64);
    put_u32(output, 24, mesh.positions.len() as u32);
    put_u32(output, 28, mesh.triangles.len() as u32);
    put_u64(output, 32, layout.positions_offset as u64);
    put_u64(output, 40, layout.triangles_offset as u64);
    put_u64(output, 48, layout.source_faces_offset as u64);
}

fn write_payload(output: &mut [u8], mesh: &IndexedTriangleMeshA0, layout: PacketLayout) {
    for (index, position) in mesh.positions.iter().flatten().enumerate() {
        output[layout.positions_offset + index * 8..layout.positions_offset + index * 8 + 8]
            .copy_from_slice(&position.to_le_bytes());
    }
    for (triangle_index, triangle) in mesh.triangles.iter().enumerate() {
        for (vertex_index, vertex) in triangle.iter().enumerate() {
            put_u32(
                output,
                layout.triangles_offset + triangle_index * 12 + vertex_index * 4,
                *vertex,
            );
        }
        if layout.source_faces_offset != 0 {
            put_u32(
                output,
                layout.source_faces_offset + triangle_index * 4,
                mesh.source_faces.as_ref().expect("validated source faces")[triangle_index],
            );
        }
    }
}

fn align_eight(value: usize) -> usize {
    (value + 7) & !7
}

fn put_u16(output: &mut [u8], offset: usize, value: u16) {
    output[offset..offset + 2].copy_from_slice(&value.to_le_bytes());
}

fn put_u32(output: &mut [u8], offset: usize, value: u32) {
    output[offset..offset + 4].copy_from_slice(&value.to_le_bytes());
}

fn put_u64(output: &mut [u8], offset: usize, value: u64) {
    output[offset..offset + 8].copy_from_slice(&value.to_le_bytes());
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn canonical_triangle_packet_has_expected_header() {
        let packet = encode_indexed_triangle_mesh_a0_packet(&IndexedTriangleMeshA0 {
            positions: vec![[0.0, 0.0, 0.0], [1.0, 0.0, 0.0], [0.0, 1.0, 0.0]],
            triangles: vec![[0, 1, 2]],
            source_faces: None,
        })
        .expect("valid packet");
        assert_eq!(&packet[..8], MAGIC);
        assert_eq!(packet.len(), 152);
        assert_eq!(&packet[16..24], &(152_u64).to_le_bytes());
    }
}

File: python/geometer/_indexed_mesh_packet_a0.py
"""Canonical indexed-triangle-mesh packet A0 codec."""

from __future__ import annotations

import math
import struct
from dataclasses import dataclass
from typing import Sequence


INDEXED_TRIANGLE_MESH_MEDIA_TYPE = "application/vnd.wavenumber.geometer.indexed-triangle-mesh"
INDEXED_TRIANGLE_MESH_PACKET_FORMAT = "geometry.indexed_triangle_mesh.packet.a0"

_MAGIC = b"GMIMSH01"
_HEADER_BYTES = 64
_HAS_SOURCE_FACES = 1
_MAX_PACKET_BYTES = 268_435_456
_MAX_VERTICES = 2_000_000
_MAX_TRIANGLES = 2_000_000
_UNSPECIFIED_SOURCE_FACE = 0xFFFF_FFFF


class IndexedTriangleMeshPacketError(ValueError):
    """The indexed-mesh value or packet violates its A0 contract."""


@dataclass(frozen=True, slots=True)
class IndexedTriangleMeshA0:
    """Flat XYZ millimeter coordinates, triangle indices, and optional source faces."""

    positions: Sequence[float]
    indices: Sequence[int]
    source_faces: Sequence[int] | None = None


def encode_indexed_triangle_mesh_a0_packet(mesh: IndexedTriangleMeshA0) -> bytes:
    """Encode a canonical ``geometry.indexed_triangle_mesh.packet.a0`` packet."""

    vertex_count, triangle_count, positions, indices, source_faces = _validated_mesh(mesh)
    has_source_faces = source_faces is not None and any(value != _UNSPECIFIED_SOURCE_FACE for value in source_faces)
    positions_offset = _HEADER_BYTES
    triangles_offset = positions_offset + vertex_count * 24
    triangles_end = triangles_offset + triangle_count * 12
    source_faces_offset = _align_eight(triangles_end) if has_source_faces else 0
    payload_end = source_faces_offset + triangle_count * 4 if has_source_faces else triangles_end
    packet_bytes = _align_eight(payload_end)
    if packet_bytes > _MAX_PACKET_BYTES:
        _fail("indexed-mesh packet exceeds 268 MiB")

    output = bytearray(packet_bytes)
    struct.pack_into(
        "<8sHHIQIIQQQQ",
        output,
        0,
        _MAGIC,
        1,
        _HEADER_BYTES,
        _HAS_SOURCE_FACES if has_source_faces else 0,
        packet_bytes,
        vertex_count,
        triangle_count,
        positions_offset,
        triangles_offset,
        source_faces_offset,
        0,
    )
    struct.pack_into(f"<{len(positions)}d", output, positions_offset, *positions)
    struct.pack_into(f"<{len(indices)}I", output, triangles_offset, *indices)
    if has_source_faces and source_faces is not None:
        struct.pack_into(f"<{len(source_faces)}I", output, source_faces_offset, *source_faces)
    return bytes(output)


def _validated_mesh(
    mesh: IndexedTriangleMeshA0,
) -> tuple[int, int, tuple[float, ...], tuple[int, ...], tuple[int, ...] | None]:
    if len(mesh.positions) < 9 or len(mesh.positions) % 3:
        _fail("positions must contain at least three complete XYZ vertices")
    if len(mesh.indices) < 3 or len(mesh.indices) % 3:
        _fail("indices must contain at least one complete triangle")
    vertex_count = len(mesh.positions) // 3
    triangle_count = len(mesh.indices) // 3
    if vertex_count > _MAX_VERTICES or triangle_count > _MAX_TRIANGLES:
        _fail("indexed mesh exceeds its governed vertex or triangle limit")
    if mesh.source_faces is not None and len(mesh.source_faces) != triangle_count:
        _fail("source_faces must contain exactly one value per triangle")

    positions = tuple(float(value) for value in mesh.positions)
    if not all(math.isfinite(value) for value in positions):
        _fail("positions must contain only finite numbers")
    indices = tuple(_uint32(value, "triangle index") for value in mesh.indices)
    if any(value >= vertex_count for value in indices):
        _fail("triangle index is outside the vertex table")
    for offset in range(0, len(indices), 3):
        if len(set(indices[offset : offset + 3])) != 3:
            _fail("each triangle must reference three distinct vertex indices")

    source_faces = (
        None if mesh.source_faces is None else tuple(_uint32(value, "source face") for value in mesh.source_faces)
    )
    return vertex_count, triangle_count, positions, indices, source_faces


def decode_indexed_triangle_mesh_a0_packet(packet: bytes | bytearray | memoryview) -> IndexedTriangleMeshA0:
    """Strictly decode and validate an indexed-triangle-mesh A0 packet."""

    data = memoryview(packet).cast("B")
    if len(data) < _HEADER_BYTES or len(data) > _MAX_PACKET_BYTES:
        _fail("indexed-mesh packet length is outside its governed bounds")
    (
        magic,
        version,
        header_bytes,
        flags,
        packet_bytes,
        vertex_count,
        triangle_count,
        positions_offset,
        triangles_offset,
        source_faces_offset,
        reserved,
    ) = struct.unpack_from("<8sHHIQIIQQQQ", data)
    has_source_faces = bool(flags & _HAS_SOURCE_FACES)
    triangles_end = triangles_offset + triangle_count * 12
    payload_end = source_faces_offset + triangle_count * 4 if has_source_faces else triangles_end
    if (
        magic != _MAGIC
        or version != 1
        or header_bytes != _HEADER_BYTES
        or flags & ~_HAS_SOURCE_FACES
        or packet_bytes != len(data)
        or not 0 < vertex_count <= _MAX_VERTICES
        or not 0 < triangle_count <= _MAX_TRIANGLES
        or positions_offset != _HEADER_BYTES
        or triangles_offset != positions_offset + vertex_count * 24
        or source_faces_offset != (_align_eight(triangles_end) if has_source_faces else 0)
        or _align_eight(payload_end) != len(data)
        or reserved != 0
    ):
        _fail("indexed-mesh packet header or table layout is invalid")
    padding_begin = triangles_end
    padding_end = source_faces_offset if has_source_faces else len(data)
    if any(data[padding_begin:padding_end]) or any(data[payload_end:]):
        _fail("indexed-mesh alignment or reserved bytes must be zero")

    positions = struct.unpack_from(f"<{vertex_count * 3}d", data, positions_offset)
    if not all(math.isfinite(value) for value in positions):
        _fail("indexed-mesh position is non-finite")
    indices = struct.unpack_from(f"<{triangle_count * 3}I", data, triangles_offset)
    if any(value >= vertex_count for value in indices):
        _fail("triangle index is outside the vertex table")
    for offset in range(0, len(indices), 3):
        if len(set(indices[offset : offset + 3])) != 3:
            _fail("triangle contains repeated vertex indices")
    source_faces = struct.unpack_from(f"<{triangle_count}I", data, source_faces_offset) if has_source_faces else None
    return IndexedTriangleMeshA0(positions=positions, indices=indices, source_faces=source_faces)


def _align_eight(value: int) -> int:
    return (value + 7) & ~7


def _uint32(value: int, label: str) -> int:
    if isinstance(value, bool) or not isinstance(value, int) or not 0 <= value <= _UNSPECIFIED_SOURCE_FACE:
        _fail(f"{label} must be an unsigned 32-bit integer")
    return value


def _fail(message: str) -> None:
    raise IndexedTriangleMeshPacketError(message)
